### PR TITLE
Extract rules into //logic and //reasoner/concludable to //logic/concludable

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?
Create `//logic` package to allow rule validation at rule creation via Graql or RPC, and enable long-lived rule caching to be performed in exactly one place (in a cache in `//logic` as opposed to `//reasoner`)

We also include `//reasoner/concludable` into `//logic/concludable` which will allow us to perform long-lived rule caching in `//logic`. Finally, update protocol to introduce `LogicManager` messages, analogous to a sort of `ConceptManager`.

## What are the changes implemented in this PR?
* create `//logic`, containing `LogicManager`, and rule implementation moved from `//concept`
* move `TypeHinter` into `//logic/hinter`
* move `//reasoner/concludable` to `//logic/concludable`
* create a single representation of `Rule` which combines the old `Rule` concept and reasoner `Implication`
* update tests to match the new structure

### CODE REVIEW

File: logic/LogicManager.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        this.conceptMgr = conceptMgr;
+        this.typeHinter = typeHinter;
+    }
+
+    public Rule putRule(String label, Conjunction<? extends Pattern> when, ThingVariable<?> then) {
+        final RuleStructure vertex = graphMgr.schema().getRule(label);
+        if (vertex != null) vertex.delete();
+        Rule rule = Rule.of(conceptMgr, graphMgr, typeHinter, label, when, then);
+
+        // TODO detect negated cycles in the rule graph after inserting this rule, requiring type hints
```
Comment (flyingsilverfin) [inline logic/LogicManager.java:46]: will be where kasper can slot in

File: logic/Rule.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    private final Set<Label> positiveWhenTypeHints;
+    private final Set<Label> negativeWhenTypeHints;
+    private final Set<Label> positiveThenTypeHints;
+    private final Set<HeadConcludable<?, ?>> head;
+    private final Set<ConjunctionConcludable<?, ?>> body;
+
+    private Rule(ConceptManager conceptMgr, RuleStructure structure, TypeHinter typeHinter) {
+        this.conceptMgr = conceptMgr;
+        this.structure = structure;
+        // TODO we should merge `when` and `then`, then compute type hints, then copy them into the `then`
```
Comment (flyingsilverfin) [inline logic/Rule.java:64]: will be where kasper can do work

File: logic/Rule.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        this.head = createHead(this.then, this.when.variables());
+        this.body = ConjunctionConcludable.of(this.when);
+    }
+
+    private Rule(GraphManager graphMgr, ConceptManager conceptMgr, TypeHinter typeHinter, String label,
+                 graql.lang.pattern.Conjunction<? extends Pattern> when, ThingVariable<?> then) {
+        graql.lang.pattern.schema.Rule.validate(label, when, then);
+        this.conceptMgr = conceptMgr;
+        this.structure = graphMgr.schema().create(label, when, then);
+        // TODO we should merge `when` and `then`, then compute type hints, then copy them into the `then`
```
Comment (flyingsilverfin) [inline logic/Rule.java:82]: will be where kasper can do work

File: logic/Rule.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        negativeWhenTypeHints = negativeTypeHints(this.when);
+        positiveThenTypeHints = positiveTypeHints(this.then);
+        validateRuleSatisfiable();
+
+        this.head = createHead(this.then, this.when.variables());
+        this.body = ConjunctionConcludable.of(this.when);
+    }
+
+    private Set<Label> positiveTypeHints(Conjunction conjunction) {
+        // TODO
```
Comment (flyingsilverfin) [inline logic/Rule.java:99]: also for kasper to do

File: logic/Rule.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        return Conjunction.create(conjunction.normalise().patterns().get(0));
+    }
+
+    private Set<Constraint> thenConstraints(ThingVariable<?> thenVariable) {
+        return VariableRegistry.createFromThings(list(thenVariable)).variables().stream().flatMap(
+                variable -> variable.constraints().stream()).collect(Collectors.toSet());
+    }
+
+    private void validateRuleSatisfiable() {
+        // TODO check that the rule has a set of satisfiable type hints. We may want to use the stream of combinations of types
```
Comment (flyingsilverfin) [inline logic/Rule.java:186]: also for kasper

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**